### DISCUSSION TITLE
Saving data when user closes browser unexpectedly

### DISCUSSION THREAD

**Discussion by wuaham** [jspsych/jsPsych#1298] [Q&A]
Hello,

I wish to save data whenever participants leave or closes the browser. I called on_close call back in jsPsych.init but it is now working if I leave/close the browser unexpectedly.

Do I need to add any plugin in my script for this please?

Thank you.

**Comment by becky-gilbert** [comment]
Hi @wuaham, no you shouldn't need to add any plugins for this. You should just be able to save your data in the `on_close` function in `jsPsych.init`, as in the example shown in the [`on_close` function documentation](https://www.jspsych.org/overview/callbacks/#on_close). So I'm not sure why this isn't working. Would you mind sharing your code please?

**Reply by wuaham** [reply]
Many thanks for the reply. My code is a bit long (around 700 lines). I have a for loop where I repeated a  randomized task couple of times but used exactly the same on_close to save data. Schematic code look like as follows:


         /*Schematic code*/
	     ...
	     ...
	     for (i = 0; i < 4; i++) {
		...
		...
		for (j = 0; j < 2; j++) {
			...
			...
			var test_procedure = {
			timeline: [fixation, test, questionmark],
			timeline_variables: test_stimuli,
			randomize_order: true,
			repetitions: 3
 			}                    
			timeline.push(hide_cursor);
			timeline.push(test_procedure);
			timeline.push(show_cursor);
                }
            }

              /*Saving data*/
            var data_save = {
                type: "html-keyboard-response",
                stimulus: "Your data will now be saved.<br /><br />" +
                        "Press any key to continue.",
                on_finish: function () {
                    saveData("experiment_data-" + start_time, jsPsych.data.get().csv());
                }
            };
            timeline.push(data_save);

           jsPsych.init({
                timeline: timeline,
                show_progress_bar: true,
                preload_images: images,
                on_close: function(){
                    saveData("experiment_data-" + start_time, jsPsych.data.get().csv());
                    }
            })

saveData function take two parameters and save data onto the server.

The entire experiment is running fine and saves data if participant completes the experiment. My concern is: if participants do not complete the experiment and leave/close the browser unexpectedly I wish to save whatever is completed.

I suspect that on_finish in data_save (var in the timeline above) contradicts/override something that prevents invoking on_close (in jsPsych.init) to save data when user closes browser unexpectedly. But if I remove the on_finish (in data_save) then how I will save data if user completes the task (as on_close is standby only to save data in unusual situation)? I may be in catch 22!

Any idea would be highly appreciated.

Thanks

**Reply by wuaham** [reply]
I am still waiting if anyone could help with any idea. Thank you.

**Reply by becky-gilbert** [reply]
Hi @wuaham, very sorry for the slow response.

You could try checking that the experiment's on_close function is being called by adding an alert inside the function:
```javascript
alert("The experiment isn't finished. Are you sure you want to leave the page?");
```
Another thing you could try is calling the function to get the CSV data first, to make sure that that function completes before the save data function is called. It might be that the page is closed right after the saveData function is called, which means that jsPsych's csv function doesn't have the chance to complete.
```javascript
var data = jsPsych.data.get().csv();
saveData(..., data);
```

@jodeleeuw not sure how to debug this since closing the page also closes the developer tools. Any ideas?

**Reply by jodeleeuw** [reply]
I think the issue is that the browser may stop executing the request to the server when the page closes. If you are sending 64kB of data or less, then you can use the [Fetch keepalive](https://javascript.info/fetch-api#keepalive) method.

We should probably update the docs so that it is clear that you are limited in what you can execute in this part of the code.

**Reply by marahaslam** [reply]
Hi!  I tried to add an alert re: Becky's suggestion above, and the alert never happens.  This indicates that on_close isn't being called at all.

**Reply by jodeleeuw** [reply]
Is this using the latest version of jsPsych?

**Reply by marahaslam** [reply]
Yes, when I run an experiment using 6.2.0 and close the tab before the experiment is done, it doesn't give an alert.

```
jsPsych.init({
    /*display_element: $('#jspsych-target'),*/
    timeline: [start_the_experiment, consent_protected_timeline, end_the_experiment],
    exclusions: {audio: true},
    preload_audio: audio,
    preload_images: images_to_preload,
    show_progress_bar: false,
    /*auto_update_progress_bar: true,*/
    on_close: function(){ 
        alert("The experiment isn't finished. Are you sure you want to leave the page?");
        var filename = "ortho_data_" + subjectID + ".csv";
        saveData(filename, jsPsych.data.get().csv());
        /*jsPsych.setProgressBar(1);*/
        console.log("data saved.");
    }
});
```
